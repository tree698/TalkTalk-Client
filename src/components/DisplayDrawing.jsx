import React from 'react';
import { useNavigate } from 'react-router-dom';
import LoggedInUser from './LoggedInUser';

export default function DisplayDrawing({
  drawing,
  checkboxForDelete,
  toBeDeletedId,
}) {
  const { id, fileName, title, username, brush, description, photo } = drawing;
  const navigate = useNavigate();

  const STYLE_FOR_LOGGEDINUSER =
    'flex items-center justify-center shrink-0 mt-0md:mt-3';
  const STYLE_FOR_AVATAR =
    'w-5 md:w-6 lg:w-8 h-5 md:h-6 lg:h-8 rounded-full mr-0 md:mr-2';
  const STYLE_FOR_USERNAME = 'hidden md:block text-sm';

  return (
    <section className="flex flex-col">
      {checkboxForDelete && (
        <input
          name="delete"
          id="delete"
          onChange={() => toBeDeletedId(id)}
          type="checkbox"
          className="w-3 md:w-4 lg:w-5 h-3 md:h-4 lg:h-5 ml-4 mb-1 border border-lightGray"
        />
      )}
      <li
        onClick={() =>
          navigate('/talk', {
            state: { id, fileName, title, username, brush, description, photo },
          })
        }
        className="list-none rounded-xl shadow-lg cursor-pointer hover:-translate-y-1 hover:brightness-125 transition-all delay-150 duration-300 ease-in-out"
      >
        <img
          src={fileName}
          alt={title}
          className="w-full h-[247px] rounded-t-xl"
        />
        <div className="flex items-center justify-center md:flex-col gap-x-2 md:gap-x-0 p-1 md:p-2 lg:p-3">
          <p className="mb-1 md:mb-2 lg:mb-3 text-sm md:text-lg lg:text-xl font-semibold">
            {title}
          </p>
          <LoggedInUser
            photo={photo}
            username={username}
            styleForAvatar={STYLE_FOR_AVATAR}
            styleForUsername={STYLE_FOR_USERNAME}
            styleForLoggedInUser={STYLE_FOR_LOGGEDINUSER}
          />
        </div>
      </li>
    </section>
  );
}
