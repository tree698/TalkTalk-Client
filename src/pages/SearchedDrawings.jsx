import React, { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import DisplayDrawing from '../components/DisplayDrawing';
import Banner from '../components/ui/Banner';
import { paginationForMyDrawingsAndSearchedDrawings } from '../config';
import { useApiContext } from '../context/ApiContext';

export default function SearchedDrawings() {
  const { limit, offset } = paginationForMyDrawingsAndSearchedDrawings;
  const { workService } = useApiContext();
  const {
    state: { searchTerm },
  } = useLocation();

  const [drawings, setDrawings] = useState([]);
  const [error, setError] = useState();

  useEffect(() => {
    searchTerm &&
      workService
        .searchWorks(limit, offset, searchTerm)
        .then((drawings) => setDrawings((prev) => [...drawings]))
        .catch((error) => setError((prev) => error.toString()));
  }, [searchTerm, workService]);

  return (
    <section className="flex-1">
      <Banner text={error} />
      <ul className="max-w-6xl mx-auto grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-4 p-4">
        {drawings.length ? (
          drawings.map((drawing) => (
            <DisplayDrawing key={drawing.id} drawing={drawing} />
          ))
        ) : (
          <Banner text="No matched drawings" />
        )}
      </ul>
    </section>
  );
}
