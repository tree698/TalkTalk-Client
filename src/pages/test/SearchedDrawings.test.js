import { Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withApiContext, withRouter } from '../../test/utils';
import SearchedDrawings from '../SearchedDrawings';
import DisplayDrawing from '../../components/DisplayDrawing';
import { fakeDrawings } from '../../test/drawingData';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

jest.mock('../../components/DisplayDrawing');
jest.mock('../../components/ui/Banner');

describe('SearchedDrawings', () => {
  const fakeContext = {
    workService: {
      searchWorks: jest.fn(),
    },
  };

  afterEach(() => fakeContext.workService.searchWorks.mockReset());

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<SearchedDrawings />} />, {
          state: {
            searchTerm: 'test',
          },
        }),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('display searched drawings', async () => {
    fakeContext.workService.searchWorks.mockImplementation(() =>
      Promise.resolve(fakeDrawings)
    );

    DisplayDrawing.mockImplementation((fakeDrawing) => (
      <pre>{JSON.stringify(fakeDrawing)}</pre>
    ));

    await act(async () => {
      render(
        withApiContext(
          withRouter(<Route path="/" element={<SearchedDrawings />} />, {
            state: {
              searchTerm: 'test',
            },
          }),
          fakeContext
        )
      );
    });

    expect(
      screen.getByText(JSON.stringify({ drawing: fakeDrawings[0] }))
    ).toBeInTheDocument();
  });
});
