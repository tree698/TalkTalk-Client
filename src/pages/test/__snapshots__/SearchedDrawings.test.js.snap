// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchedDrawings renders correctly 1`] = `
<section
  className="flex flex-col items-center"
>
  <ul
    className="w-full grid grid-cols-1 md:grid-cols-3 lg:grid-cols-4 gap-1 md:gap-2 lg:gap-3 mb-6 md:mb-8 lg:mb-10 px-3 md:px-0"
  />
  <div
    className="mt-8 md:mt-10 lg:mt-12 text-xs md:text-sm lg:text-base"
  >
    <svg
      className="text-lg md:text-base lg:text-xl mx-auto"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 632H136v-39.9l138.5-164.3 150.1 178L658.1 489 888 761.6V792zm0-129.8L664.2 396.8c-3.2-3.8-9-3.8-12.2 0L424.6 666.4l-144-170.7c-3.2-3.8-9-3.8-12.2 0L136 652.7V232h752v430.2zM304 456a88 88 0 1 0 0-176 88 88 0 0 0 0 176zm0-116c15.5 0 28 12.5 28 28s-12.5 28-28 28-28-12.5-28-28 12.5-28 28-28z"
      />
    </svg>
  </div>
</section>
`;
